## Q1 LDA

In [18]:
import math
import numpy as np

In [19]:
#sample for class 1
x1=[(5, 3), (3, 5), (3, 4), (4, 7), (5, 6)]

In [20]:
#sample for class 2
x2=[(9, 10), (7, 7), (8, 5), (7, 2), (10, 8)]

In [21]:
#class mean
mu1=np.mean(x1,axis=0)
mu2=np.mean(x2,axis=0)

In [22]:
#Covariance matrix of the class
s1=np.cov(x1,rowvar=False)
s2=np.cov(x2,rowvar=False)

In [47]:
#Within-class scatter matrix
sw=s1+s2
sw

array([[ 2.7 ,  2.65],
       [ 2.65, 11.8 ]])

In [48]:
#Between-class scatter matrix:
sb=np.dot((np.array([mu1-mu2]).T),(np.array([mu1-mu2])))
sb

array([[17.64,  5.88],
       [ 5.88,  1.96]])

In [50]:
#compute the LDA projection
inverse_sw = np.linalg.inv(sw)
inverse_sw

array([[ 0.47508807, -0.10669351],
       [-0.10669351,  0.10870659]])

In [51]:
inverse_sw_by_sb=np.dot(inverse_sw,sb)
inverse_sw_by_sb

array([[ 7.75319577,  2.58439859],
       [-1.24287871, -0.4142929 ]])

In [53]:
# get the projection vector
w,v=np.linalg.eig(inverse_sw_by_sb)

In [56]:
#The eigenvalues
w

array([7.33890287e+00, 1.11022302e-16])

In [57]:
#The normalized (unit “length”) eigenvectors
v

array([[ 0.98739355, -0.31622777],
       [-0.15828446,  0.9486833 ]])

In [63]:
#The optimal projection is the one that given maximum λ = J(w)
v[:,0]

array([ 0.98739355, -0.15828446])

In [64]:
w[0]

7.338902868646196

##### Perform LDA and find out the optimal projection vector (normalized to unit length) is [ 0.98739355, -0.15828446]

##### and its corresponding eigenvalue is 7.338902